WalletWasabi.Fluent/ViewModels/CoinSelection/SelectCoinsDialogViewModel.cs

```diff
@@ -0,0 +1,160 @@
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Reactive;
+using System.Reactive.Disposables;
+using System.Reactive.Linq;
+using DynamicData;
+using NBitcoin;
+using ReactiveUI;
+using WalletWasabi.Blockchain.Analysis.Clustering;
+using WalletWasabi.Blockchain.TransactionOutputs;
+using WalletWasabi.Fluent.Extensions;
+using WalletWasabi.Fluent.Helpers;
+using WalletWasabi.Fluent.ViewModels.CoinSelection.Core;
+using WalletWasabi.Fluent.ViewModels.Dialogs.Base;
+using WalletWasabi.Fluent.ViewModels.Wallets;
+using WalletWasabi.Fluent.ViewModels.Wallets.Send;
+using ICoin = WalletWasabi.Fluent.ViewModels.CoinSelection.Core.ICoin;
+
+namespace WalletWasabi.Fluent.ViewModels.CoinSelection;
+
+[NavigationMetaData(
+	Title = "Select Coins",
+	Caption = "",
+	IconName = "wallet_action_send",
+	NavBarPosition = NavBarPosition.None,
+	Searchable = false,
+	NavigationTarget = NavigationTarget.DialogScreen)]
+public partial class SelectCoinsDialogViewModel : DialogViewModelBase<IEnumerable<SmartCoin>>
+{
+	private readonly IObservable<Unit> _balanceChanged;
+	private readonly SmartLabel _transactionLabels;
+	private readonly IEnumerable<SmartCoin> _predefinedCoins;
+	private readonly WalletViewModel _walletViewModel;
+	private readonly TransactionInfo _transactionInfo;
+
+	[AutoNotify] private string _searchFilter = "";
+	[AutoNotify] private ReactiveCommand<Unit, Unit> _clearCoinSelectionCommand = ReactiveCommand.Create(() => { });
+	[AutoNotify] private IObservable<bool> _enoughSelected = Observable.Return(false);
+	[AutoNotify] private IObservable<bool> _isSelectionBadlyChosen = Observable.Return(false);
+	[AutoNotify] private LabelBasedCoinSelectionViewModel? _labelBasedSelection;
+	[AutoNotify] private IObservable<Money> _remainingAmount = Observable.Return(Money.Zero);
+	[AutoNotify] private ReactiveCommand<Unit, Unit> _selectAllCoinsCommand = ReactiveCommand.Create(() => { });
+	[AutoNotify] private ReactiveCommand<Unit, Unit> _selectPrivateCoinsCommand = ReactiveCommand.Create(() => { });
+	[AutoNotify] private IObservable<Money> _selectedAmount = Observable.Return(Money.Zero);
+	[AutoNotify] private IObservable<int> _selectedCount = Observable.Return(0);
+	[AutoNotify] private IObservable<Money> _requiredAmount = Observable.Return(Money.Zero);
+	[AutoNotify] private ReactiveCommand<Unit, Unit> _selectPredefinedCoinsCommand = ReactiveCommand.Create(() => { });
+	[AutoNotify] private IObservable<string> _summaryText = Observable.Return("");
```

Comment:
This extremely smells.

Rather init things in the ctor... and also create a disposable and dispose it when the dialog is popped from the navigation stack.

```c#
protected override void OnNavigatedFrom(bool isInHistory)
{
	base.OnNavigatedFrom(isInHistory);

	if (!isInHistory)
	{
		// call dispose here
	}
}
```

Reply:
This is the problem of not using the disposable pattern correctly.

Reply:
> also create a disposable and dispose it when the dialog is popped from the navigation stack.
> 
> ```cs
> protected override void OnNavigatedFrom(bool isInHistory)
> {
> 	base.OnNavigatedFrom(isInHistory);
> 
> 	if (!isInHistory)
> 	{
> 		// call dispose here
> 	}
> }
> ```

IMHO, this is not necessary, because the disposable will be disposed here:

https://github.com/zkSNACKs/WalletWasabi/blob/1e673d9028b438f30de36953c9138ec2aab7ccc7/WalletWasabi.Fluent/ViewModels/Navigation/RoutableViewModel.cs#L61

Reply:
You but the difference is the lifetime of the disposable. While this one lives during the ViewModel is being displayed, the one that you would create in the ctor would leave until the ViewModel is needed. Now it works, but in the future you might be able to navigate away from the coin selection page and then back.

> This is one of the problems of not using the Disposable pattern correctly. The ViewModel should be destroyed when it's no longer used, but we keep it alive instead.

Yepp I see the problem. But this is not that easy because there are ViewModels that lifetimes are equal to the software (and being displayed more than once), and there the others that live until they are in the navigation stack.
Make ViewModels disposable? Mehh... problematic doing it manually and automatizing it is also problematic.

##

Do something with this because it is not good. It is a huge hack to eliminate the null warnings. The code should be written as objects cannot be null, and not to make them auto-notifiable and set i.e. commands that do nothing.

Reply:
I have revamped initialization. I think it's better now and **fixed** presumably.

Just as a note, *ctor initialization* was my first attempt, but then I remembered that `OnNavigatedTo` is used for initialization. To be honest, I still don't know 100% sure if OnNavigatedTo is meant to be used for initialization, but it looks so. It provides a `CompositeDisposable` that I supposed to be disposed by the navigation mechanism after leaving the VM. It's used in many VMs, like:

https://github.com/zkSNACKs/WalletWasabi/blob/1e673d9028b438f30de36953c9138ec2aab7ccc7/WalletWasabi.Fluent/ViewModels/Wallets/Home/History/Details/CoinJoinsDetailsViewModel.cs#L50

In order to use it, I needed to delay initialization until `OnNavigatedTo` has been called. This is terrible design for immutability among other problems.

I'd like to ditch this approach (*deferred init*) in favor of *ctor init*. 

This would involve.
1. Not using OnNavigatedTo to perform initializations, ignoring the provided `CompositeDisposable`.
2. Managing disposal ourselves in `OnNavigatedFrom`.

I hope you agree with this.